cdap-tms/src/main/java/co/cask/cdap/messaging/client/ClientMessagingService.java

```diff
@@ -0,0 +1,511 @@
+  private static final Type TOPIC_PROPERTY_TYPE = new TypeToken<Map<String, String>>() { }.getType();
+  private static final Type TOPIC_LIST_TYPE = new TypeToken<List<String>>() { }.getType();
+
+  private final EndpointStrategy endpointStrategy;
+
+  @VisibleForTesting
+  @Inject
+  public ClientMessagingService(DiscoveryServiceClient discoveryServiceClient) {
+    this.endpointStrategy = new RandomEndpointStrategy(
+      discoveryServiceClient.discover(Constants.Service.MESSAGING_SERVICE));
+  }
+
+  @Override
+  public void createTopic(TopicMetadata topicMetadata) throws TopicAlreadyExistsException, IOException {
+    TopicId topicId = topicMetadata.getTopicId();
+
+    HttpRequest request = HttpRequest
+      .put(createURL(createTopicPath(topicId)))
+      .withBody(GSON.toJson(topicMetadata.getProperties()))
+      .build();
+    HttpResponse response = HttpRequests.execute(request, HTTP_REQUEST_CONFIG);
+
+    if (response.getResponseCode() == HttpURLConnection.HTTP_CONFLICT) {
+      throw new TopicAlreadyExistsException(topicId);
+    }
+    handleError(response, "Failed to create topic " + topicId);
+  }
+
+  @Override
+  public void updateTopic(TopicMetadata topicMetadata) throws TopicNotFoundException, IOException {
+    TopicId topicId = topicMetadata.getTopicId();
+
+    HttpRequest request = HttpRequest
+      .put(createURL(createTopicPath(topicId) + "/properties"))
+      .withBody(GSON.toJson(topicMetadata.getProperties()))
+      .build();
+    HttpResponse response = HttpRequests.execute(request, HTTP_REQUEST_CONFIG);
+
+    if (response.getResponseCode() == HttpURLConnection.HTTP_NOT_FOUND) {
+      throw new TopicNotFoundException(topicId);
+    }
+    handleError(response, "Failed to update topic " + topicId);
+  }
+
+  @Override
+  public void deleteTopic(TopicId topicId) throws TopicNotFoundException, IOException {
+    HttpRequest request = HttpRequest
+      .delete(createURL(createTopicPath(topicId)))
+      .build();
+    HttpResponse response = HttpRequests.execute(request, HTTP_REQUEST_CONFIG);
+
+    if (response.getResponseCode() == HttpURLConnection.HTTP_NOT_FOUND) {
+      throw new TopicNotFoundException(topicId);
+    }
+    handleError(response, "Failed to update topic " + topicId);
+  }
+
+  @Override
+  public TopicMetadata getTopic(TopicId topicId) throws TopicNotFoundException, IOException {
+    HttpRequest request = HttpRequest
+      .get(createURL(createTopicPath(topicId)))
+      .build();
+    HttpResponse response = HttpRequests.execute(request, HTTP_REQUEST_CONFIG);
+
+    if (response.getResponseCode() == HttpURLConnection.HTTP_NOT_FOUND) {
+      throw new TopicNotFoundException(topicId);
+    }
+    handleError(response, "Failed to update topic " + topicId);
+
+    Map<String, String> properties = GSON.fromJson(response.getResponseBodyAsString(), TOPIC_PROPERTY_TYPE);
+    return new TopicMetadata(topicId, properties);
```

Comment:
With the current logic, it handle all errors first. This is similar to early termination pattern (i.e. `if (!valid) throw something` first). The code would be less indented and easier to follow (all errors has been handled first, instead of interleaving error/ok/error).